apache/superset
# perf: refactor SIP-68 db migrations with INSERT SELECT FROM

### SUMMARY

This is another take on #19406 and #19416. The goal is to rewrite the bulk of loading + rewriting operations from Python to native SQL statements by utilizing `INSERT SELECT FROM`.

The whole migration happens in 3 steps:

1. Copy `tables`, `table_columns`, `sql_metrics` to the new `sl_xxx` tables using INSERT SELECT FROM.
2. Copy model associations to relationship tables.
3. Run post processing on `sl_datastes` and `sl_columns` to
   - Tuck additional metadata columns (`verbose_name`, `d3format`, etc) in the original tables under `extra_json`
   - Add conditional quotes to physical table and column expressions
   - Apply dataset level `is_managed_externally` and `external_url` to columns

Comparing to #19406, this PR was able to cut down the migration time for our Superset instance from 7 hours to 1 hour while retaining the full functionality.

While working on the migration, I also noticed a couple of shortcomings in the original implementation, including problems in the shadow-writing logic. Most notably:

1. Some long text columns should use `MediumText` for MySQL
2. Columns are not automatically deleted when they are unlinked from a table or dataset (e.g. when you modify or remove a column from a dataset)
3. Letting physical table and dataset point to the same columns could be problematic as we wouldn't want users who updated column properties in a dataset to automatically propagate that change to the related table. Tables and datasets may have different ownerships and different users may create different physical datasets for different tables. One of the original goal of SIP-68 was to change the mapping between physical datasets and database table/views to an n:1 relationship. It only makes sense if column definitions in each dataset are independent as well.
4. Dataset does not know which database it belongs to unless being joined with a table. However, a dataset does not always have an associated table---for example, we can have virtual queries like "SELECT 1;" that does not populate data from any tables. Or, the table name extraction for a complex SQL statement could fail (especially when it has Jinja syntax).

Instead of creating multiple db migrations to address these issues, it'd be much easier for end users if we just skip the original migration (drop data for who already migrated) and bundle everything in another migration:

1. Datasets and columns now reuse uuids from the old tables. This makes filling up the association tables easier. It also makes it possible to import a Superset export from an old Superset instance that hasn't upgrade to the new models, even after the old models have been removed.
2. `created_on`, `change_on`, `created_by` and `changed_by` are also copied over when appropriate.
3. In MySQL, some text columns (`expression`, `description`, etc) are now `MediumText`. This is consistent with current column types in the old tables.
4. Since old and new entities can now be linked via `uuid`, `sqlatable_id` on `NewDataset` is removed. However, there is no 1:1 mapping of `NewTable` in the old models so they acquires uuids. In order to link `NewTable` with associated columns, `sqlatable_id` is added to `NewTable` during the db migration, but dropped once migration is complete.
5. Refactored shadow-writing hooks to reduce duplicate code and more appropriately handle column deletion.
6. Removed column fetching for tables. We used to fetch table metadata (column types, etc) from datasources during db migration, as well as in shadowing writing, but this has a major risk---external datasources can be unreliably slow. Users could fail the migration or saving edits to a datasource if the datasource is temporarily down. We should not let this unreliable process block the critical path of basic functionalities. I added a `Table.sync_columns` method, but in the future this should be consolidated with `get_virtual_table_metadata` and `get_physical_table_metadata`.
7. `dataset.owners` are also added (copying PR #19487).

#### Known issues and next steps

1. The table columns for `NewTable` will only be copied from the dataset when its being created. We'd need another mechanism to sync table columns. Either during re-populating after `fetch_metadata` or in an async script. Since the Table model is not used anywhere yet, I think it's okay to leave this in another followup PR.
2. I added a [`sync_columns`](https://github.com/apache/superset/blob/03295342428faff6958f7a4a1181ac69bb88c6fd/superset/tables/models.py#L122) method for tables. It's not currently in use and does more or less the same as `SqlaTable.fetch_metadata`. I think it's okay to keep this somewhat-duplicate not-in-active-use method as `SqlaTable` will be remove eventually.
3. In this PR, `NewTable.columns` will be updated only when a **physical** `SqlaTable` is first synced to the new models and a `NewTable` is created. Updating columns in a `SqlaTable` dataset will NOT update `SqlaTable.columns` as we consider table columns and dataset columns to be separate and the metadata syncing should only happen in one direction---that is from tables to datasets, not the other way around.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

<img width="600" alt="image" src="https://user-images.githubusercontent.com/335541/161164424-b12d34ce-b7ab-4a6b-a6a1-b0ca2b25f649.png">


### TESTING INSTRUCTIONS

To test the migration

```shell
superset db downgrade 2ed890b36b94 && SHOW_PROGRESS=1 superset db upgrade a9422eeaae74
```

A couple of queries to verify the correctness of the migrated data:

```sql
select
  sl_tables.name as table_name,
  sl_datasets.name as dataset_name,
  sl_datasets.expression
from sl_datasets
  join sl_dataset_tables on sl_datasets.id = sl_dataset_tables.dataset_id
  join sl_tables on sl_tables.id = sl_dataset_tables.table_id
limit 5;
```

```sql
select id, name, extra_json, changed_by_fk
from sl_columns
where
  changed_by_fk is not null
  and extra_json like "%verbose_name%"
limit 10;
```

```sql
select
  c.id, c.name, c.expression, c.uuid, c.created_on
from sl_columns as c join sl_dataset_columns a
  on a.column_id = c.id
where a.dataset_id = <your id>
order by created_on desc limit 10;
```

### ADDITIONAL INFORMATION


- [x] Has associated issue: SIP-68, #14909 and #17543
